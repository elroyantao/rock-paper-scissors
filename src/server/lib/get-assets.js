import { WDS_PORT } from '../../shared/config'

const isProd = process.env.NODE_ENV === 'production'


export function getScripts() {
  const scripts = [
    'js/bundle.js'
  ]
  return scripts.map((script) => {
    return `<script src="${isProd ? '/public/' : `http://localhost:${WDS_PORT}/public/`}${script}"></script>`
  })
}
