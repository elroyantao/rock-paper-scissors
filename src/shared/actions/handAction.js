export const setPlayerHand = (playerHand) => ({
  type: 'SET_PLAYER_HAND',
  playerHand
})

export const setComputerHand = (computerHand) => ({
  type: 'SET_COMPUTER_HAND',
  computerHand
})

export const clearHand = () => ({
  type: 'CLEAR_HANDS'
})
