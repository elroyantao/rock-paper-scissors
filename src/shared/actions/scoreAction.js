export const playerWins = () => ({
  type: 'PLAYER_WINS'
})

export const computerWins = () => ({
  type: 'COMPUTER_WINS'
})

export const itsATie = () => ({
  type: 'ITS_A_TIE'
})

export const resetScore = () => ({
  type: 'RESET_SCORE'
})
