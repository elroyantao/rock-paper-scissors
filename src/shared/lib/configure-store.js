import { createStore, applyMiddleware, compose } from 'redux'
import thunkMiddleware from 'redux-thunk'

import reducers from './combine-reducers'

const finalCreateStore = compose(
  applyMiddleware(thunkMiddleware),
  typeof window === 'object' && typeof window.devToolsExtension !== 'undefined' ?
    window.devToolsExtension() : (f) => f
)(createStore)

export default function configureStore(initialState = {}) {
  const store = finalCreateStore(
    reducers,
    initialState
  )
  return store
}
