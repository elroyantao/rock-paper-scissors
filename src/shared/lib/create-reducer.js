export default function (initialState, handlers) {
  return function reducer(state, action) {
    const newState = (state === undefined) ? initialState : state

    if (Object.prototype.hasOwnProperty.call(handlers, action.type)) {
      return handlers[action.type](newState, action)
    }

    return newState
  }
}
