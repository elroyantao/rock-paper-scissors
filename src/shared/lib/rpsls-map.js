export const ROCK = 0
export const SPOCK = 1
export const PAPER = 2
export const LIZARD = 3
export const SCISSORS = 4

export const classic = [ROCK, PAPER, SCISSORS]
export const advanced = [ROCK, PAPER, SCISSORS, LIZARD, SPOCK]
