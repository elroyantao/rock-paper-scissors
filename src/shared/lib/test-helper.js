import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

const renderComponentHelper = (Component, mountOptions = {}) => {
  return ({ children, ...props } = {}) => {
    const wrapper = shallow(
      <Component {...props}>{ children }</Component>,
      mountOptions
    )

    const component = {
      wrapper,
      instance: wrapper.instance(),
      getTree: () => toJson(wrapper),
      getTreeFor: (selectedWrapper) => toJson(selectedWrapper)
    }

    return component
  }
}

export default renderComponentHelper
