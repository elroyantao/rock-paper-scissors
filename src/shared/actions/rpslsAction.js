import { setPlayerHand, setComputerHand } from './handAction'
import { playerWins, computerWins, itsATie } from './scoreAction'

export function findWinner(player1, player2) {
  if ((player2 + 1) % 5 === player1 || (player2 + 2) % 5 === player1) return 1
  else if (player2 === player1) return 0
  return 2
}

export const getComputerHand = (availableHands) => {
  return new Promise((resolve) => {
    setTimeout(() => {
      const hand = availableHands[Math.floor(Math.random() * availableHands.length)]
      resolve(hand)
    }, 1000)
  })
}

export const applyHand = (playerHand) => async (dispatch, getState) => {
  const state = getState()
  dispatch(setPlayerHand(playerHand))
  const computerHand = await getComputerHand(state.level.hands)
  dispatch(setComputerHand(computerHand))
  const winner = findWinner(playerHand, computerHand)
  if (!winner) dispatch(itsATie())
  else if (winner === 1) dispatch(playerWins())
  else if (winner === 2) dispatch(computerWins())
}
