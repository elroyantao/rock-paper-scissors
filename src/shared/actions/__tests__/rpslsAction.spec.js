import * as actions from '../rpslsAction'
import * as rpsls from '../../lib/rpsls-map'
import { setPlayerHand, setComputerHand } from '../handAction'
import { playerWins, computerWins, itsATie } from '../scoreAction'

jest.mock('../handAction', () => ({
  setPlayerHand: jest.fn(),
  setComputerHand: jest.fn()
}))
jest.mock('../scoreAction', () => ({
  itsATie: jest.fn(),
  playerWins: jest.fn(),
  computerWins: jest.fn()
}))

const rpslsArray = rpsls.advanced

describe('rpslsAction', () => {
  describe('findWinner', () => {
    it('when player1 is Rock and player2 is Rock 1 the response should be 0 / TIE', () => {
      expect(actions.findWinner(rpsls.ROCK, rpsls.ROCK)).toBe(0)
    })
    it('when player1 is Rock and player2 is Scissors 1 the response should be 1 / player1', () => {
      expect(actions.findWinner(rpsls.ROCK, rpsls.SCISSORS)).toBe(1)
    })
    it('when player1 is LIZARD and player2 is SPOCK 1 the response should be 1 / player1', () => {
      expect(actions.findWinner(rpsls.LIZARD, rpsls.SPOCK)).toBe(1)
    })
    it('when player1 is PAPER and player2 is SCISSORS 1 the response should be 2 / player2', () => {
      expect(actions.findWinner(rpsls.PAPER, rpsls.SCISSORS)).toBe(2)
    })
    rpslsArray.forEach((player1) => {
      rpslsArray.forEach((player2) => {
        it(`snapshot the winners when player1 is ${Object.keys(rpsls)[player1]} and player2 is ${Object.keys(rpsls)[player2]}`, () => {
          expect(actions.findWinner(player1, player2))
            .toMatchSnapshot()
        })
      })
    })
  })
  describe('getComputerHand', () => {
    beforeEach(() => jest.clearAllMocks())
    it('should give select a random hand from the availableHands and despond after 1s', () => {
      expect.assertions(1)
      return actions.getComputerHand(rpslsArray).then((hand) => {
        expect(rpslsArray.includes(hand)).toBe(true)
      })
    })
    it('should call the setTimeout to 1s', () => {
      jest.useFakeTimers()
      actions.getComputerHand(rpslsArray)
      expect(setTimeout.mock.calls.length).toBe(1)
      expect(setTimeout.mock.calls[0][1]).toBe(1000)
      jest.useRealTimers()
    })
  })

  describe('applyHand', () => {
    // jest.useFakeTimers()
    const dispatch = jest.fn(val => val)
    const getState = jest.fn(() => ({
      level: {
        hands: rpslsArray
      }
    }))
    const action = actions.applyHand(2)
    beforeAll(() => {
      actions.default.__Rewire__('getComputerHand', jest.fn(() => 2))
    })
    afterAll(() => {
      actions.default.__ResetDependency__('getComputerHand')
    })
    beforeEach(() => jest.clearAllMocks())

    it('should call dispatch 3 times and getState 1 time', async () => {
      await action(dispatch, getState)
      expect(dispatch).toHaveBeenCalledTimes(3)
      expect(getState).toHaveBeenCalledTimes(1)
    })
    it('should call setPlayerHand and setComputerHand 1 time', async () => {
      await action(dispatch, getState)
      expect(setPlayerHand).toHaveBeenCalledTimes(1)
      expect(setComputerHand).toHaveBeenCalledTimes(1)
    })
    it('should call getComputerHand 1 time', async () => {
      await action(dispatch, getState)
      expect(actions.default.__get__('getComputerHand')).toHaveBeenCalledTimes(1)
    })
    it('should call itsATie when both select the same', async () => {
      const action = actions.applyHand(2)
      await action(dispatch, getState)
      expect(itsATie).toHaveBeenCalledTimes(1)
      expect(playerWins).not.toHaveBeenCalled()
      expect(computerWins).not.toHaveBeenCalled()
    })
    it('should call playerWins when player selects Scissors and computer selects paper', async () => {
      const action = actions.applyHand(4)
      await action(dispatch, getState)
      expect(playerWins).toHaveBeenCalledTimes(1)
      expect(itsATie).not.toHaveBeenCalled()
      expect(computerWins).not.toHaveBeenCalled()
    })
    it('should call computerWins when player selects rock and computer selects paper', async () => {
      const action = actions.applyHand(0)
      await action(dispatch, getState)
      expect(computerWins).toHaveBeenCalledTimes(1)
      expect(itsATie).not.toHaveBeenCalled()
      expect(playerWins).not.toHaveBeenCalled()
    })
  })
})
