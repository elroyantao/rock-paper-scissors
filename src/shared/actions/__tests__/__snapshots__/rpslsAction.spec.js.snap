// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rpslsAction findWinner snapshot the winners when player1 is LIZARD and player2 is LIZARD 1`] = `0`;

exports[`rpslsAction findWinner snapshot the winners when player1 is LIZARD and player2 is PAPER 1`] = `1`;

exports[`rpslsAction findWinner snapshot the winners when player1 is LIZARD and player2 is ROCK 1`] = `2`;

exports[`rpslsAction findWinner snapshot the winners when player1 is LIZARD and player2 is SCISSORS 1`] = `2`;

exports[`rpslsAction findWinner snapshot the winners when player1 is LIZARD and player2 is SPOCK 1`] = `1`;

exports[`rpslsAction findWinner snapshot the winners when player1 is PAPER and player2 is LIZARD 1`] = `2`;

exports[`rpslsAction findWinner snapshot the winners when player1 is PAPER and player2 is PAPER 1`] = `0`;

exports[`rpslsAction findWinner snapshot the winners when player1 is PAPER and player2 is ROCK 1`] = `1`;

exports[`rpslsAction findWinner snapshot the winners when player1 is PAPER and player2 is SCISSORS 1`] = `2`;

exports[`rpslsAction findWinner snapshot the winners when player1 is PAPER and player2 is SPOCK 1`] = `1`;

exports[`rpslsAction findWinner snapshot the winners when player1 is ROCK and player2 is LIZARD 1`] = `1`;

exports[`rpslsAction findWinner snapshot the winners when player1 is ROCK and player2 is PAPER 1`] = `2`;

exports[`rpslsAction findWinner snapshot the winners when player1 is ROCK and player2 is ROCK 1`] = `0`;

exports[`rpslsAction findWinner snapshot the winners when player1 is ROCK and player2 is SCISSORS 1`] = `1`;

exports[`rpslsAction findWinner snapshot the winners when player1 is ROCK and player2 is SPOCK 1`] = `2`;

exports[`rpslsAction findWinner snapshot the winners when player1 is SCISSORS and player2 is LIZARD 1`] = `1`;

exports[`rpslsAction findWinner snapshot the winners when player1 is SCISSORS and player2 is PAPER 1`] = `1`;

exports[`rpslsAction findWinner snapshot the winners when player1 is SCISSORS and player2 is ROCK 1`] = `2`;

exports[`rpslsAction findWinner snapshot the winners when player1 is SCISSORS and player2 is SCISSORS 1`] = `0`;

exports[`rpslsAction findWinner snapshot the winners when player1 is SCISSORS and player2 is SPOCK 1`] = `2`;

exports[`rpslsAction findWinner snapshot the winners when player1 is SPOCK and player2 is LIZARD 1`] = `2`;

exports[`rpslsAction findWinner snapshot the winners when player1 is SPOCK and player2 is PAPER 1`] = `2`;

exports[`rpslsAction findWinner snapshot the winners when player1 is SPOCK and player2 is ROCK 1`] = `1`;

exports[`rpslsAction findWinner snapshot the winners when player1 is SPOCK and player2 is SCISSORS 1`] = `1`;

exports[`rpslsAction findWinner snapshot the winners when player1 is SPOCK and player2 is SPOCK 1`] = `0`;
